### PR TITLE
Register Intel distributed Backend (`XCCL`) in PyTorch distributed package

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
### Motivation:

As design illustrated in Intel distributed support RFC https://github.com/pytorch/pytorch/issues/141741, two sections are needed to enable intel distributed backend (`XCCL`) support in PyTorch.
1. Intel GPU distributed Backend integration in PyTorch `torch-xpu-ops`.
2. **Intel distributed Backend register in PyTorch distributed package**. This PR is to contribute section 2 change.  

### Example:
Here is a simple example of using spawn to launch XCCL backend and perform allreduce on XPU tensors.
```
import os
import torch
import torch.distributed as dist
import torch.multiprocessing as mp

def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '29500'
    dist.init_process_group(rank=rank, world_size=world_size)

def cleanup():
    dist.destroy_process_group()

def run_allreduce(rank, world_size):
    setup(rank, world_size)
    device = torch.device('xpu:{}'.format(rank))
    x = torch.randn([2, 2], device=device)
    dist.all_reduce(x)
    cleanup()

if __name__ == '__main__':
    world_size = 2
    mp.spawn(run_allreduce, args=(world_size,), nprocs=world_size, join=True)
```

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Comment by gujinghui** [comment]
Address the RFC https://github.com/pytorch/pytorch/issues/141741

**Review by kwen2501 — APPROVED** [review]
LGTM.
Ideally we don't need the pybind as part of the registration. But it comes from the limitation of how we create backend today -- at python level. We can host the code as is, while taking some time to figure out a better solution.

**Comment by gujinghui** [comment]
> LGTM. Ideally we don't need the pybind as part of the registration. But it comes from the limitation of how we create backend today -- at python level. We can host the code as is, while taking some time to figure out a better solution.

Sounds great! Let's keep it as is, then, start to think how to refine the registration process. Thanks.

**Review by gujinghui — APPROVED** [review]

**Comment by guangyey** [comment]
@zhangxiaoli73 you need to add xccl to https://github.com/pytorch/pytorch/blob/f675f644fd0fc4202527671422e4072d9ecf6622/torch/_C/_distributed_c10d.pyi#L304

**Comment by zhangxiaoli73** [comment]
> @zhangxiaoli73 you need to add xccl to
> 
> https://github.com/pytorch/pytorch/blob/f675f644fd0fc4202527671422e4072d9ecf6622/torch/_C/_distributed_c10d.pyi#L298-L306

Got you. This will cause a lint check failure. Fix done.

**Comment by guangyey** [comment]
@zhangxiaoli73 you need to add the full name `torch.distributed.distributed_c10d.is_xccl_available` to https://github.com/pytorch/pytorch/blob/ce22a01e11f9fce76efa482def197fd5eae9954a/docs/source/distributed.rst?plain=1#L197 to fix the doc issue.

**Comment by guangyey** [comment]
"rebase to the latest viable/strict to fix XPU build issue."
@pytorchbot rebase

**Comment by zhangxiaoli73** [comment]
"rebase to the latest viable/strict to fix XPU build issue."
@pytorchbot rebase

File: docs/source/conf.py [hunk 1]
```diff
@@ -519,6 +519,7 @@
     "is_initialized",
     "is_mpi_available",
     "is_nccl_available",
+    "is_xccl_available",
```
Comment (albanD) [inline docs/source/conf.py:522]: Adding new entries to this file is not ok.
Given you added the function to the .rst below, this should not be needed unless something is wrong with the doc setup there.

**Review by albanD — CHANGES_REQUESTED** [review]

File: docs/source/conf.py (hunk 1 above)
Comment (guangyey) [reply docs/source/conf.py:522]: @zhangxiaoli73, please add the full name `torch.distributed.distributed_c10d.is_xccl_available` to the doc and remove `is_xccl_available` from this list.

Comment (zhangxiaoli73) [reply docs/source/conf.py]: @albanD Thanks for your reminder. Revoked the change in this file. Please help review.

**Review by albanD — APPROVED** [review]
LGTM, thanks for the update!

**Pull request closed without merging**